### ISSUE TITLE
False positive: invalid-arity when inner binding shadows a local function name

### ISSUE THREAD

**Issue by yuhan0** [clj-kondo/clj-kondo#2854]
- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**version**
clj-kondo v2026.05.25

**platform**
native

**editor**
NA

**problem**

When a `let` or `letfn`-bound function name is shadowed by an inner let binding or fn parameter, clj-kondo wrongly uses the _outer_ function's constraints to lint its calls, resulting in false positives:

```clojure
(let [foo (fn ^long [x] (count x)) ;; unused binding
      foo vector]
  (foo 1 2 3)        ;; => error: foo is called with 3 args but expects 1
  (reverse (foo :a)) ;; => warning: Expected: seqable collection, received: long.
  )
```

This also happens with letfn bindings: 
```clojure
(letfn [(f ([] 0) ([_] (f)))
        (dmap [f xs]
          (map f ;; => error: f is called with 2 args but expects 0 or 1
            xs (next xs)))]
  (dmap str [1 2 3 4]))
```

Both the above examples are false positives - they evaluate without errors. 
Also note that this is specific to locally bindings: global vars are shadowed correctly (ie. `(defn f []) (let [f inc] (f 1))` does not produce lint warnings)

**repro**
```clojure
(ns repro)

(letfn [(f [x y] (+ x y))]
  (let [f (partial f 1)]
    (f 2)))
```

```bash
$ clj-kondo --lint repro.clj
repro.clj:5:5: error: f is called with 1 arg but expects 2
linting took 47ms, errors: 1, warnings: 0
```

**config**
NA

**expected behavior**

No errors should be reported.

[ I had Claude investigate and produce a reasonable-looking fix in impl.analyzer that passes tests - I'm not confident to say I fully understand the surrounding context, but if that's alright I can submit a PR. ]

**Comment by borkdude** [comment]
Nice find!

**Comment by borkdude** [comment]
About the PR: if you don't understand the PR, the burden is on me to figure out what Claude wrote. I might as wel fix it myself then I guess :)

**Comment by yuhan0** [comment]
Here was the fix for reference anyway:
```diff
modified   src/clj_kondo/impl/analyzer.clj
@@ -503,7 +503,9 @@ defn
                 arity analyzed-arg-vec arglist-str arg-vec]
          return-tag :ret
          arg-tags :args} (analyze-fn-arity ctx body)
-        ctx (ctx-with-bindings ctx arg-bindings)
+        ctx (cond-> (ctx-with-bindings ctx arg-bindings)
+              (seq arg-bindings)
+              (update :arities #(apply dissoc % (keys arg-bindings))))
         ctx (assoc ctx
                    :recur-arity arity
                    :top-level? false)
@@ -969,7 +971,7 @@ defn
                                  ;; since functions cannot be destructured
                                  (assoc arities binding-val (assoc arity
                                                                    :types types-by-arity))
-                                 arities)]
+                                 (dissoc arities binding-val))]
               (when cbu-collector
                 (swap! cbu-collector conj
                        (and (= 2 (some-> pair-tracker deref))
@@ -1041,7 +1043,7 @@ defn
                               (analyze-children
                                (-> ctx
                                    (ctx-with-bindings analyzed-bindings)
-                                   (update :arities merge arities)
+                                   (assoc :arities arities)
                                    (assoc :let-parent let-parent))
                                let-body
                                false)))]
```

**Comment by borkdude** [comment]
@yuhan0 Looks very similar to the PR I have up now
